### PR TITLE
[fix] Follow pep8 guidelines

### PR DESCRIPTION
**Pull request by lyz-code**
Hi, I'm trying to contribute some code to interact with the history of Taskwarrior to get the active time of the tasks.

But I've seen that there were some PEP8 and other guideline warnings and I've decided to fix them.

I hope you're OK with it :)

### CODE REVIEW

**Comment by lyz-code** [comment]
@robgolding @tbabej It seems that travis is failing because it's trying to clone an old repo
```bash
git clone --recursive https://git.tasktools.org/scm/tm/task.git
```

Maybe it should point to `https://git.tasktools.org/tbabej/task`

**Comment by lyz-code** [comment]
@robgolding how can I rerun the tests?

**Comment by robgolding** [comment]
@lyz-code if you merge `master` into this branch Travis should re-run the build. However there are some indentation/linting styles in this PR that are a little unusual so wouldn't get merged. Most of the stuff is an improvement, though!

**Comment by lyz-code** [comment]
@robgolding It seem's to be passing the tests now.

Which indentation/linting is unusual? I've followed the guidelines of the [python-mode](https://github.com/python-mode/python-mode) vim plugin :S 

Just tell me which lines are and I'll revert them so it can be merged.

Thanks for your time!

:)

**Comment by lyz-code** [comment]
Hi @robgolding, I guess you are busy (I see it in your commit frequency).

Whenever you can I'd be glad if you could tell me what do you see odd. I see you have a lot of experience and I'd like to learn :)

Furthermore, I'm also working in a feature of `tasklib` to parse the history and work with times natively. I'd like to know your oppinion on that soon.

Thanks for your work

**Comment by robgolding** [comment]
Hey @lyz-code! Sorry for being unattentive here. I'll do a code review with examples where the formatting is a bit unusual, and my preferred alternative. Note that this is personal opinion, and my interpretation of PEP8! 😄

File: tasklib/lazy.py [hunk 1]
```diff
@@ -172,19 +172,19 @@ def issuperset(self, other):
...
 
     def difference(self, other):
         return LazyUUIDTaskSet(self._tw,
-            self._uuids - set(t['uuid'] for t in other))
+                               self._uuids - set(t['uuid'] for t in other))
 
     def symmetric_difference(self, other):
         return LazyUUIDTaskSet(self._tw,
-            self._uuids ^ set(t['uuid'] for t in other))
+                               self._uuids ^ set(t['uuid'] for t in other))
```
Comment (robgolding) [inline tasklib/lazy.py]: Same as above for all these 4 cases.

File: tasklib/task.py [hunk 2]
```diff
@@ -333,7 +333,8 @@ def serialize_depends(self, cur_dependencies):
 
     def delete(self):
         if not self.saved:
-            raise Task.NotSaved("Task needs to be saved before it can be deleted")
+            raise Task.NotSaved(
+                "Task needs to be saved before it can be deleted")
```
Comment (robgolding) [inline tasklib/task.py]: 2 things to point out here for ultimate code compliance 😄 

1. Strings should be single-quoted where possible (that is, unless the string itself contains a single quote).
2. If you start a new line after an opening parenthesis `(`, close the parentheses on their own line too (and remember the trailing commas!).

So:

```python
raise Task.NotSaved(
    'Task needs to be saved before it can be deleted',
)
```

Note that if the string is too long and causes the line to be longer than 80 characters, use implicit string concatenation:

```python
raise Task.NotSaved(
    'Task needs to be saved before it can be deleted, because '
    ' of some reason that makes the string really long!',
)

File: tasklib/task.py [hunk 3]
```diff
@@ -382,7 +385,8 @@ def stop(self):
 
     def done(self):
         if not self.saved:
-            raise Task.NotSaved("Task needs to be saved before it can be completed")
+            raise Task.NotSaved(
+                "Task needs to be saved before it can be completed")
```
Comment (robgolding) [inline tasklib/task.py]: Same as above in these 3 cases.

File: tasklib/tests.py [hunk 4]
```diff
@@ -124,7 +126,8 @@ def test_filtering_by_attribute(self):
 
         # Assert that the correct tasks are returned
         high_priority_task = self.tw.tasks.get(priority="H")
-        self.assertEqual(high_priority_task['description'], "high priority task")
+        self.assertEqual(high_priority_task['description'],
+                         "high priority task")
```
Comment (robgolding) [inline tasklib/tests.py]: Again, it's always good to break after the opening parenthesis in method calls like this, and especially in tests with `assertEqual` I find it much easier to see what's going on (also prefer single quotes as before):

```python
self.assertEqual(
    high_priority_task['description'],
    'high priority task',
)
```

File: tasklib/tests.py [hunk 5]
```diff
@@ -277,7 +280,7 @@ class TaskTest(TasklibTest):
 
     def test_create_unsaved_task(self):
         # Make sure a new task is not saved unless explicitly called for
-        t = Task(self.tw, description="test task")
+        Task(self.tw, description="test task")
```
Comment (robgolding) [inline tasklib/tests.py]: Use single quotes here.

File: tasklib/tests.py [hunk 6]
```diff
@@ -327,9 +330,9 @@ def test_add_completed_task(self):
 
     def test_add_multiple_completed_tasks(self):
         t1 = Task(self.tw, description="test1", status="completed",
-                 end=datetime.datetime.now())
+                  end=datetime.datetime.now())
```
Comment (robgolding) [inline tasklib/tests.py]: I know it's verbose, but putting each argument on its own line if they won't all fit on a single one makes things really clear, in my opinion. It also means you can add/remove arguments while keeping a very clear diff in Git.

```python
t1 = Task(
    self.tw,
    description="test1",
    status="completed",
    end=datetime.datetime.now(),
)
```

File: tasklib/tests.py [hunk 7]
```diff
@@ -327,9 +330,9 @@ def test_add_completed_task(self):
...
         t2 = Task(self.tw, description="test2", status="completed",
-                 end=datetime.datetime.now())
+                  end=datetime.datetime.now())
```
Comment (robgolding) [inline tasklib/tests.py]: Same as above here.

File: tasklib/tests.py [hunk 8]
```diff
@@ -777,7 +780,7 @@ def test_deserializer_returning_empty_value_for_empty_string(self):
         # Test that any deserializer returns empty value when passed ''
         t = Task(self.tw)
         deserializers = [getattr(t, deserializer_name) for deserializer_name in
-                        filter(lambda x: x.startswith('deserialize_'), dir(t))]
+                         filter(lambda x: x.startswith('deserialize_'), dir(t))]
```
Comment (robgolding) [inline tasklib/tests.py]: List comprehensions can be "exploded" onto multiple lines to make them much clearer, like so:

```python
deserializers = [
    getattr(t, deserializer_name)
    for deserializer_name in filter(
        lambda x: x.startswith('deserialize_'),
        dir(t),
    )
]
```

In this case I also split the `filter` call onto multiple lines, but that's not strictly necessary as it'd fit on a single line.

File: tasklib/tests.py [hunk 9]
```diff
@@ -818,7 +821,8 @@ def test_return_all_from_executed_command(self):
     def test_return_all_from_failed_executed_command(self):
         Task(self.tw, description="test task", tags=['test']).save()
         out, err, rc = self.tw.execute_command(['countinvalid'],
-            return_all=True, allow_failure=False)
+                                               return_all=True,
+                                               allow_failure=False)
```
Comment (robgolding) [inline tasklib/tests.py]: I'd probably opt for each argument on a new line here:

```python
out, err, rc = self.tw.execute_command(
    ['countinvalid'],
    return_all=True,
    allow_failure=False,
)
```

File: tasklib/tests.py [hunk 10]
```diff
@@ -839,16 +843,17 @@ class TaskFromHookTest(TasklibTest):
         '"status":"pending",'
         '"uuid":"81305335-0237-49ff-8e87-b3cdc2369cec"}')
 
-    input_modify_data = six.StringIO(input_add_data.getvalue() + '\n' +
+    input_modify_data = six.StringIO(
+        input_add_data.getvalue() + '\n' +
```
Comment (robgolding) [inline tasklib/tests.py]: In this case, we could use `str.join` to make the code a bit clearer:

```python
input_modify_data = six.StringIO(
    '\n'.join([
        input_add_data.getvalue(),
        (
            '{"description":"Buy some milk finally",'
            '"entry":"20141118T050231Z",'
            '"status":"completed",'
            '"uuid":"a360fc44-315c-4366-b70c-ea7e7520b749"}'
        ),
)

File: tasklib/tests.py [hunk 11]
```diff
@@ -839,16 +843,17 @@ class TaskFromHookTest(TasklibTest):
...
         '"entry":"20141118T050231Z",'
         '"status":"completed",'
         '"uuid":"a360fc44-315c-4366-b70c-ea7e7520b749"}')
 
     exported_raw_data = (
         '{"project":"Home",'
-         '"due":"20150101T232323Z",'
-         '"description":"test task"}')
+        '"due":"20150101T232323Z",'
+        '"description":"test task"}')
```
Comment (robgolding) [inline tasklib/tests.py]: The closing `)` should be on a new line here, since the opening one is on its own line.

File: tasklib/tests.py [hunk 12]
```diff
@@ -875,8 +880,9 @@ def test_setting_up_from_modified_hook_input(self):
 
     def test_export_data(self):
         t = Task(self.tw, description="test task",
-            project="Home",
-            due=pytz.utc.localize(datetime.datetime(2015,1,1,23,23,23)))
+                 project="Home",
+                 due=pytz.utc.localize(datetime.datetime(2015, 1, 1,
+                                                         23, 23, 23)))
```
Comment (robgolding) [inline tasklib/tests.py]: Instead of breaking up the parentheses that are "smaller" here, it's best to go back to the "larger" ones and break them up first (as it often means you can keep related things on the same line). In this case, we can also put each argument to `Task` on its own line, and use single quotes too:

```python
t = Task(
    self.tw,
    description='test task',
    project='Home',
    due=pytz.utc.localize(
        datetime.datetime(2015, 1, 1, 23, 23, 23),
    ),
)
```

Note: I don't know for sure if the `localize()` line would all fit inside 80 characters, so I broke it up just to show how it can be done.

File: tasklib/tests.py [hunk 13]
```diff
@@ -1020,7 +1028,7 @@ def test_complex_eoy_conversion(self):
             hour=23,
             minute=59,
             second=59
-            )) - datetime.timedelta(0,4 * 30 * 86400)
+            )) - datetime.timedelta(0, 4 * 30 * 86400)
```
Comment (robgolding) [inline tasklib/tests.py]: This line can be de-dented one level, so that the opening and closing parentheses are at the same indentation level. Also the `second=59` line should have a trailing comma.

File: tasklib/tests.py [hunk 14]
```diff
@@ -1111,40 +1120,43 @@ def test_non_unicode_task(self):
...
-                         list(k for k in self.sample))
+        self.assertEqual(list(key for key in self.view),
+                         list(key for key in self.sample))
 
         # Assert the view is correct after modification
         self.sample['new'] = 'value'
-        self.assertEqual(list(k for k in self.view),
-                         list(k for k in self.sample))
+        self.assertEqual(list(key for key in self.view),
+                         list(key for key in self.sample))
```
Comment (robgolding) [inline tasklib/tests.py]: I'd probably opt to separate these arguments to `assertEqual` out onto their own lines, as they are a little bit hard to parse as-is with the `list()` calls in there as well.

File: tasklib/tests.py [hunk 15]
```diff
@@ -1350,8 +1362,10 @@ def test_symmetric_difference(self):
 
         assert taskset ^ lazyset == set([self.task1, self.task3])
         assert lazyset ^ taskset == set([self.task1, self.task3])
-        assert taskset.symmetric_difference(lazyset) == set([self.task1, self.task3])
-        assert lazyset.symmetric_difference(taskset) == set([self.task1, self.task3])
+        assert taskset.symmetric_difference(lazyset) == set([self.task1,
+                                                             self.task3])
+        assert lazyset.symmetric_difference(taskset) == set([self.task1,
+                                                             self.task3])
```
Comment (robgolding) [inline tasklib/tests.py]: This should be using `self.assertEqual` really, which would make the lines a lot easier to format 😄

File: setup.py [hunk 16]
```diff
@@ -34,6 +34,6 @@
         "Programming Language :: Python :: 3.3",
         'License :: OSI Approved :: BSD License',
         'Topic :: Software Development :: Libraries :: Python Modules',
-        'Intended Audience :: Developers',
+        'Intended Audience :: Developers'
```
[the remark this answers is not in the document]
Comment (jamatute) [reply setup.py]: Fixed in 8f4fabf

File: tasklib/backends.py [hunk 17]
```diff
@@ -214,7 +215,8 @@ def format_description(self, task):
         if self.version < self.VERSION_2_4_0:
             return task._data['description']
         else:
-            return six.u("description:'{0}'").format(task._data['description'] or '')
+            return six.u("description:'{0}'").format(task._data['description']
+                                                     or '')
```
[the remark this answers is not in the document]
Comment (jamatute) [reply tasklib/backends.py]: Fixed in 8f4fabf

File: tasklib/lazy.py [hunk 18]
```diff
@@ -129,7 +129,7 @@ def __isub__(self, other):
 
     def __rsub__(self, other):
         return LazyUUIDTaskSet(self._tw,
-            set(t['uuid'] for t in other) - self._uuids)
+                               set(t['uuid'] for t in other) - self._uuids)
```
[the remark this answers is not in the document]
Comment (jamatute) [reply tasklib/lazy.py]: Fixed in 8f4fabf

File: tasklib/lazy.py (hunk 1 above)
Comment (jamatute) [reply tasklib/lazy.py]: Fixed in 8f4fabf

File: tasklib/task.py (hunk 2 above)
Comment (jamatute) [reply tasklib/task.py]: Fixed in 8f4fabf

File: tasklib/task.py (hunk 3 above)
Comment (jamatute) [reply tasklib/task.py]: Fixed in 8f4fabf

File: tasklib/tests.py (hunk 4 above)
Comment (jamatute) [reply tasklib/tests.py]: Fixed in 8f4fabf

File: tasklib/tests.py (hunk 5 above)
Comment (jamatute) [reply tasklib/tests.py]: Fixed in 8f4fabf

File: tasklib/tests.py (hunk 6 above)
Comment (jamatute) [reply tasklib/tests.py]: Fixed in 8f4fabf

File: tasklib/tests.py (hunk 7 above)
Comment (jamatute) [reply tasklib/tests.py]: Fixed in 8f4fabf

File: tasklib/tests.py (hunk 8 above)
Comment (jamatute) [reply tasklib/tests.py]: Fixed in 8f4fabf

File: tasklib/tests.py (hunk 9 above)
Comment (jamatute) [reply tasklib/tests.py]: Fixed in 8f4fabf

File: tasklib/tests.py (hunk 10 above)
Comment (jamatute) [reply tasklib/tests.py]: I specially liked this example, thanks!

Fixed in 8f4fabf

File: tasklib/tests.py (hunk 11 above)
Comment (jamatute) [reply tasklib/tests.py]: Fixed in 8f4fabf

File: tasklib/tests.py (hunk 12 above)
Comment (jamatute) [reply tasklib/tests.py]: Fixed in 8f4fabf

File: tasklib/tests.py (hunk 13 above)
Comment (jamatute) [reply tasklib/tests.py]: I didn't really catch what you meant with

> can be de-dented one level, so the opening...

But I think I fixed it in 8f4fabf

File: tasklib/tests.py (hunk 14 above)
Comment (jamatute) [reply tasklib/tests.py]: Fixed in 8f4fabf

File: tasklib/tests.py (hunk 15 above)
Comment (jamatute) [reply tasklib/tests.py]: I think I fixed it in 8f4fabf.

Thank you so much for taking the time of making the review. It served **a
lot**, my contributions from now on will incorporate it ^^

I've also tried to apply the guidelines in the affected files, but I did it
kind of quick so some one might have slipped

Thanks again @robgolding

**Comment by robgolding** [comment]
No problem, it looks great now! And all the tests pass, so let’s merge 👍

File: tasklib/task.py (hunk 2 above)
Comment (tbabej) [reply tasklib/task.py]: Randomly stumbling here, I wonder @robgolding why the preference for single quotes?

Comment (robgolding) [reply tasklib/task.py]: Simply a personal preference @tbabej, which is not documented in any coding guidelines (yet!) 😄
